src/datasets/features.py

```diff
@@ -33,18 +34,60 @@
 
 
 def string_to_arrow(type_str: str) -> pa.DataType:
-    if type_str not in pa.__dict__:
+    """
+    string_to_arrow takes a string pyarrow dtype and converts it to a pyarrow.DataType.
+
+    In effect, `dt == string_to_arrow(str(dt))`
+
+    This is necessary because the datasets.Value() primitive type is constructed using a string dtype, oftentimes
+
+    Value(dtype=str(pa_type))
+
+    But Features.type (via `get_nested_type()` expects to resolve Features into a pyarrow Schema,
+        which means that each Value() must resolve into its corresponding pyarrow.DataType, which is the purpose
+        of this function.
+    """
+    timestamp_regex = re.compile(r"^timestamp\[(.*)\]$")
+    timestamp_matches = timestamp_regex.search(type_str)
+    if timestamp_matches:
+        """
+        >>> import pyarrow as pa
+        >>> str(pa.timestamp('us'))
+        'timestamp[us]'
+        >>> str(pa.timestamp('us', tz='America/New_York'))
+        'timestamp[us, tz=America/New_York]'
+        """
+        timestamp_internals = timestamp_matches.group(1)
+        internals_regex = re.compile(r"^(s|ms|us|ns),\s*tz=([a-zA-Z0-9/_+:]*)$")
+        internals_matches = internals_regex.search(timestamp_internals)
+        if timestamp_internals in ["s", "ms", "us", "ns"]:
+            return pa.timestamp(timestamp_internals)
+        elif internals_matches:
+            return pa.timestamp(internals_matches.group(1), internals_matches.group(2))
+        else:
+            raise ValueError(
+                f"{type_str} is not a validly formatted string representation of a pyarrow timestamp."
+                f"Examples include timestamp[us] or timestamp[us, tz=America/New_York]"
+                f"See: https://arrow.apache.org/docs/python/generated/pyarrow.timestamp.html#pyarrow.timestamp"
+            )
+    elif type_str == "double":
```

Comment:
this handles the fact that `str(pa.float64()) == "double"` by inverting it, which is redundant with https://github.com/huggingface/datasets/pull/1900/files#diff-54081ede051fd0a7ef65748c481cc06f90209f01bb89968747089d13a2ca052bR182

Reply:
I'm ok for having this so that the `dt == string_to_arrow(str(dt))` rule holds.
Not that in practice since Value resolves "double" to "float64", then this case shouldn't happen to get the pyarrow type of a Value type.